data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/LogicalReplicationEventProcessor.java

```diff
@@ -0,0 +1,262 @@
+package org.opensearch.dataprepper.plugins.source.rds.stream;
+
+import org.opensearch.dataprepper.buffer.common.BufferAccumulator;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.event.JacksonEvent;
+import org.opensearch.dataprepper.model.opensearch.OpenSearchBulkActions;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.source.rds.RdsSourceConfig;
+import org.opensearch.dataprepper.plugins.source.rds.converter.StreamRecordConverter;
+import org.opensearch.dataprepper.plugins.source.rds.coordination.partition.StreamPartition;
+import org.opensearch.dataprepper.plugins.source.rds.datatype.postgres.ColumnType;
+import org.opensearch.dataprepper.plugins.source.rds.model.TableMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class LogicalReplicationEventProcessor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LogicalReplicationEventProcessor.class);
+
+    static final Duration BUFFER_TIMEOUT = Duration.ofSeconds(60);
+    static final int DEFAULT_BUFFER_BATCH_SIZE = 1_000;
+
+    private final StreamPartition streamPartition;
+    private final StreamRecordConverter recordConverter;
+    private final Buffer<Record<Event>> buffer;
+    private final BufferAccumulator<Record<Event>> bufferAccumulator;
+    private final List<Event> pipelineEvents;
+
+    private long currentLsn;
+    private long currentEventTimestamp;
+
+    private Map<Long, TableMetadata> tableMetadataMap;
+
+    public LogicalReplicationEventProcessor(final StreamPartition streamPartition,
+                                            final RdsSourceConfig sourceConfig,
+                                            final Buffer<Record<Event>> buffer,
+                                            final String s3Prefix) {
+        this.streamPartition = streamPartition;
+        recordConverter = new StreamRecordConverter(s3Prefix, sourceConfig.getPartitionCount());
+        this.buffer = buffer;
+        bufferAccumulator = BufferAccumulator.create(buffer, DEFAULT_BUFFER_BATCH_SIZE, BUFFER_TIMEOUT);
+
+        tableMetadataMap = new HashMap<>();
+        pipelineEvents = new ArrayList<>();
+    }
+
+    public void process(ByteBuffer msg) {
+        // Message processing logic:
+        // If it's a BEGIN, note its LSN
+        // If it's a RELATION, update table metadata map
+        // If it's INSERT/UPDATE/DELETE, prepare events
+        // If it's a COMMIT, convert all prepared events and send to buffer
+        char messageType = (char) msg.get();
+        if (messageType == 'B') {
```

Comment:
Consider using enums instead of char for message types for type safety, readability and maintainability

Reply:
Done.